config/puma.rb

```diff
@@ -0,0 +1,21 @@
+# https://devcenter.heroku.com/articles/deploying-rails-applications-with-the-puma-web-server
+
+workers Integer(ENV['WEB_CONCURRENCY'] || 1)
```

Comment:
RE: raising mem, it sounds like increasing the available memory will allow for greater concurrency/threads which is a good thing assuming that the core resources (CPU, RAM (no swap), disk/net IO) on the host are not taxed.

I'm no expert but this is what I (think I) understand. If your code was 100% CPU bound then you want one thread per CPU (I think the ideal is two with hyper threading (HT) enabled chips). A 100% CPU bound app is usually never the case and you're going to have to do some experimenting to find out what amount of conc/threads works best for your setup/resource constraints (historically we'd be limited to physical HW, 2-8 core chips and max of ~32GB ram per host). To test, I think you want to map concurrency to the number of CPU cores. Then, start adding more threads and your performance should increase since the processor is most likely going to be waiting on disk/net IO and free to do other tasks (you'll want to be watching cpu, mem (swap), disk/net IO during these tests). I would keep adding threads until you see one of your core resources constrained (memory is swapping, cpu is > 100%, etc) or a steady degradation in performance in which case your now loosing to the excess context switching between threads, not the disk/net IO. If memory is flexible, just keep an eye on it and make sure there's enough to prevent it from having to swap.

Let us know how it goes.